text/0123-pending-code-as-storage-location-for-runtime-upgrades.md

```diff
@@ -0,0 +1,60 @@
+## Drawbacks
+
+Because the first block built with the new runtime code will move the runtime code from `:pending_code` to `:code`, the runtime code will need to be loaded. This means the runtime code will appear in the proof of validity of a parachain for the first block built with the new runtime code. Generally this  is not a problem as the runtime code is also loaded by the parachain when setting the new runtime code.
+There is still the possibility of having state that is not migrated even when following the proposal as presented by this RFC. The issue is that if the amount of data to be migrated is too big, not all of it can be migrated in one block, because either it takes more time than there is assigned for a block or parachains for example have a fixed budget for their proof of validity. To solve this issue there already exist multi-block migrations that can chunk the migration across multiple blocks. Consensus-critical data needs to be migrated in the first block to ensure that block production etc., can continue. For the other data being migrated by multi-block migrations the migrations could for example expose to the outside which keys are being migrated and should not be indexed until the migration is finished.
+
+## Testing, Security, and Privacy
+
+Testing should be straightforward and most of the existing testing should already be good enough. Extending with some checks that `:pending_code` is moved to `:code`.
+## Performance, Ergonomics, and Compatibility
+
+### Performance
+
+The performance should not be impacted besides requiring loading the runtime code in the first block being built with the new runtime code.
+
+### Ergonomics
+
+It only alters the way blocks are produced and imported after applying a runtime upgrade. This means that only nodes need to be adapted to the changes of this RFC.
+
+### Compatibility
+
+The change will require that the nodes are upgraded before the runtime starts using this feature. Otherwise they will fail to import the block build by `:pending_code`. 
+For Polkadot/Kusama this means that also the parachain nodes need to be running with a relay chain node version that supports this new feature. Otherwise the parachains will stop producing/finalizing nodes as they can not sync the relay chain any more.
+
+## Prior Art and References
+
+The [issue](https://github.com/paritytech/polkadot-sdk/issues/64) initially reported a bug that led to this RFC. It also discusses multiple solutions for the problem.
+
+## Unresolved Questions
+
+None
+
+## Future Directions and Related Material
+
+- Solve the issue of requiring loading the entire runtime code to move it into a different location by introducing a low-level `move` function. When using the `V1` trie layout every value bigger than 32 bytes is put into the db separately. This means a low level `move` function would only need to move the hash of the runtime code from `:code` to `:pending_code`.
```

Comment:
:+1:

Reply:
Without this, don't we have an even more severe risk of needing a multi block migration?

Reply:
If you want to migrate 5MIB at once, yes. However, even now I would not recommend you to do this in one migration. Also it is not a "severe risk", you just need to use a multi block migration ;) These already exist.